### PR TITLE
Translating Uncategorised Observation to Observation Statement

### PR DESCRIPTION
**Pull request by z-uddin**
* Created a ObservationStatement mapper, which consumes FHIR Observation object and uses this to fill an Observation Statement XML template.
* Unit tests parse a JSON file to a FHIR Observation object and then compares output from running map method to expected XML
* Created a DateFormat util to convert dates to yyyyMMddHHmmss
* Refactored unit tests in NarrativeStatementMapperTest to use parameterized test and argument stream for resource files

### CODE REVIEW

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/utils/DateFormatUtil.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        .toFormatter();
+
+    public static String formatDate(Date date) {
+        if (date == null) {
+            throw new EhrMapperException("Could not format date");
+        }
+
+        return DATE_TIME_FORMATTER.format(
+            date
+                .toInstant()
```
Comment (mmackiewicz1) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/utils/DateFormatUtil.java:23]: Could we move this method invocation one line up?

File: service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ObservationStatementMapperTest.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+    private static final String TEST_ID = "394559384658936";
+    private static final String INPUT_JSON_WITH_EFFECTIVE_DATE_TIME = "/ehr/mapper/observation/example-observation-resource-1.json";
+    private static final String INPUT_JSON_WITH_NULL_DATE_TIME = "/ehr/mapper/observation/example-observation-resource-2.json";
+    private static final String INPUT_JSON_WITH_EFFECTIVE_PERIOD = "/ehr/mapper/observation/example-observation-resource-3.json";
+    private static final String INPUT_JSON_WITH_ISSUED_ONLY = "/ehr/mapper/observation/example-observation-resource-4.json";
+    private static final String INPUT_JSON_WITH_NO_DATES = "/ehr/mapper/observation/example-observation-resource-5.json";
+    private static final String OUTPUT_XML_USES_EFFECTIVE_DATE_TIME = "/ehr/mapper/observation/expected-output-observation-statement-1.xml";
+    private static final String OUTPUT_XML_USES_UNK_DATE_TIME = "/ehr/mapper/observation/expected-output-observation-statement-2.xml";
+    private static final String OUTPUT_XML_USES_NESTED_COMPONENT = "/ehr/mapper/observation/expected-output-observation-statement-3.xml";
```
Comment (mmackiewicz1) [inline service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ObservationStatementMapperTest.java:38]: I noticed that "/ehr/mapper/observation/" suffix is common for all the constants. Maybe we could move it to a separate one for a better readability?

File: service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ObservationStatementMapperTest.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+public class ObservationStatementMapperTest {
+    @Mock
+    private RandomIdGeneratorService randomIdGeneratorService;
```
Comment (mmackiewicz1) [inline service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ObservationStatementMapperTest.java:28]: Instance fields are always after static ones.

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ObservationStatementMapper.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+    public String mapObservationToObservationStatement(Observation observation, boolean isNested) {
+        var observationStatementTemplateParameters = ObservationStatementTemplateParameters.builder()
+            .observationStatementId(randomIdGeneratorService.createNewId())
+            .comment(observation.getComment())
+            .issued(DateFormatUtil.formatDate(observation.getIssued()))
+            .isNested(isNested)
+            .build();
+
+        if (observation.hasEffectiveDateTimeType() && observation.getEffectiveDateTimeType().hasValue()) {
```
Comment (msolecka) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ObservationStatementMapper.java:32]: 3 returns in one function is not the best pattern, what do you think about changing it into one return and only configure it in the ifs? 

```
template = OBSERVATION_STATEMENT_EFFECTIVE_PERIOD_TEMPLATE
...

return  TemplateUtils.fillTemplate(template, observationStatementTemplateParameters);
````

but if you applied my previous comment this code would probably change

**Review by mmackiewicz1 — APPROVED** [review]

File: service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ObservationStatementMapperTest.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    @BeforeEach
+    public void setUp() {
+        when(randomIdGeneratorService.createNewId()).thenReturn(TEST_ID);
+        observationStatementMapper = new ObservationStatementMapper(randomIdGeneratorService);
+    }
+
+    @Test
+    public void When_MappingParsedObservationJsonWithEffectiveDatetime_Expect_ObservationStatementXmlOutput() throws IOException {
```
Comment (mmackiewicz1) [inline service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ObservationStatementMapperTest.java:62]: This test can be parametrised. Please take a look at "EhrExtractRequestHandlerTest" for a reference.

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ObservationStatementMapper.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+@Component
+public class ObservationStatementMapper {
+
+    private static final Mustache OBSERVATION_STATEMENT_EFFECTIVE_TIME_TEMPLATE =
+        TemplateUtils.loadTemplate("ehr_observation_statement_effective_time_template.mustache");
+    private static final String EFFECTIVE_TIME_XML = "<effectiveTime><center value=\"%s\"/></effectiveTime>";
+    private static final String EFFECTIVE_PERIOD_XML = "<effectiveTime><low value=\"%s\"/><high value=\"%s\"/></effectiveTime>";
+    private static final String EFFECTIVE_TIME_UNK_XML =  "<effectiveTime><center value=\"UNK\"/></effectiveTime>";
+
+    private final RandomIdGeneratorService randomIdGeneratorService;
```
Comment (msolecka) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ObservationStatementMapper.java:23]:
```suggestion
    private final MessageContext messageContext;
```

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ObservationStatementMapper.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+    public String mapObservationToObservationStatement(Observation observation, boolean isNested) {
+        var observationStatementTemplateParameters = ObservationStatementTemplateParameters.builder()
+            .observationStatementId(randomIdGeneratorService.createNewId())
```
Comment (msolecka) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ObservationStatementMapper.java:27]:
```suggestion
            .observationStatementId(messageContext.getIdMapper().getOrNew(ResourceType.Observation, observation.getId()))
```

File: service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/NarrativeStatementMapperTest.java [hunk 8]
```diff
@@ -59,58 +62,30 @@ public void setUp() {
...
     }
 
-    @Test
-    public void When_MappingParsedObservationJsonWithEffectiveDatetime_Expect_NarrativeStatementXmlOutput() throws IOException {
-        expectedOutputMessage = ResourceTestFileUtils.getFileContent(OUTPUT_XML_USES_EFFECTIVE_DATE_TIME);
-
-        var jsonInput = ResourceTestFileUtils.getFileContent(INPUT_JSON_WITH_EFFECTIVE_DATE_TIME);
+    @ParameterizedTest
+    @MethodSource("resourceFileParams")
+    public void When_MappingObservationJson_Expect_NarrativeStatementXmlOutput(String inputJson, String outputXml) throws IOException {
```
Comment (mmackiewicz1) [inline service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/NarrativeStatementMapperTest.java:67]: 👍

File: service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ObservationStatementMapperTest.java [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+        TimeZone.setDefault(null);
+    }
+
+    @BeforeEach
+    public void setUp() {
+        when(randomIdGeneratorService.createNewId()).thenReturn(TEST_ID);
+        observationStatementMapper = new ObservationStatementMapper(randomIdGeneratorService);
+    }
+
+    @ParameterizedTest
```
Comment (mmackiewicz1) [inline service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ObservationStatementMapperTest.java:65]: 👍

**Review by mmackiewicz1 — APPROVED** [review]

**Review by msolecka — APPROVED** [review]

**Pull request merged**